docs/versioning.md

```diff
@@ -0,0 +1,101 @@
+3. A consumer of a given ONNX model MAY consume an updated ONNX file provided there are one or more breaking changes in the new ONNX file's IR version, referenced operator versions, or model version.
+
+Per the rules of SemVer 2.0, during the initial development of ONNX, we use a MAJOR version of 0 for both the IR version and operator version and will only increment the MINOR version in the face of either a breaking change or the need to stabiize a version for specific engineering purposes. [ISSUE: decide how we will stabilize and archive released versions of ONNX] 
+
+### Serializing SemVer version numbers in protobuf
+
+For historical reasons, ONNX serializes the MAJOR, MINOR, and PATCH values as a bit-packed 32-bit integer; the most siginificant byte is the MAJOR component, the second most significant byte is the MINOR component, the least significant two bytes are the PATCH component. 
+
+For example, 1.2.3 is represented as 0x01020003.
+
+The pre-release and build metadata are represented as ISSUE: are we supporting these or not, and if so, on what entity types? 
+
+
+
+
+
+## IR versioning
+
+Changes to the file format or graph model semantics version atomically. Breaking changes to the format or semantics of the ONNX specification require an increment of the MAJOR version.  Non-breaking format or semantic changes that introduce new functionality require an increment of the MINOR version. Non-breaking changes to the specification that simply clarify spec ambiguities require an increment of the PATCH version.  
+
+The ONNX IR format adheres to the versioning guidelines defined in the [Updating a Message Type](https://developers.google.com/protocol-buffers/docs/proto3#updating) section of the proto3 specification.  
+
+As a general principle, implementations SHOULD be robust in the face of missing fields. However, to ensure basic interoperation, a subset of message fields will be marked as required for a given IR version and all producers MUST set these fields correctly. Required fields always are marked with the comment:
+
+    // This field MUST be present for this version of the IR.
+
+By way of example, the ModelProto.ir_version MUST be present in every model.
+
+ISSUE: decide and document how we want to handle changes to the format that are forward compatible but will cause the generated code to introduce a breaking change (e.g., changing the data type of a field from int64 to int32, int64 to double or bytes to string).  I would recommend that we never do it after we hit 1.0.0 - prior to that we should be fairly liberal with changes to size (e.g., int32<>int16, float<>double) and conservative with changes to value space (e.g., integral<>floatingpoint, string<>int, scalar<>message)
```

Comment:
A shared/centralized library (c++ and c#, for example) to wrap generated codes into internal representation, with user friendly APIs, may resolve this issue to some extent. In this way, producers or consumers will program against the shared library. Format changes do not break data compatibility but break code compatibility will be handled by the shared library itself.

Reply:
I'd advocate for adding a statement here that we guarantee forever backward-compatibility for protobuf-level deserialization. It's how you're supposed to use protobufs anyway. If we ever need to change the field type we'd introduce a new field and deprecate the old one

Reply:
First, Ke's characterization of data compat vs. code compat is spot on. They are two different things, and the vast majority of protobuf versioning rules are focused on the former and IMO ignore the latter.  If you read the [protobuf versioning rules](https://developers.google.com/protocol-buffers/docs/proto3#updating), it say it's totally OK to switch from int32 to bool, but that would break A TON of code if one were to regenerate their language bindings using protoc against the new .proto file. 

As for Ke's proposed solution, while we may elect to provide a shared library, the specification cannot mandate or really assume it will be used everywhere. We have to assume a reasonable number of people will just run protoc on the latest .proto file and have to absorb our "data compat/code incompat) changes (assuming we every let them happen).  BTW, we are some of those people so we get to live with whatever policy we apply. 

As for @dzhulgakov's suggestion of keeping a field type immutable, I'm OK with that PROVIDED we use a new name for the field. Otherwise, we get the worst of both worlds: we cause the same breaking change to consumers of the .proto file AND we add versioning complexity to consumers as the field's value can now appear in two different places based on version.

BTW, once we hit V1.0.0, I want to adopt @dzhulgakov's approach. Leading up to that, I'm more liberal, as the blast radius of a BC is still fairly contained.  That honeymoon will end soon enough, so maybe we just adopt that policy now (with my addendum of not reusing the same field NAME as well as field NUMBER).  

Dmytro, are you OK with this?

Reply:
I actually didn't realize that Protobuf is permissive with some type changes (I'm somewhat biased with Thrift's experience). In that case expanding the type to broader type if we need it should be permissible.

For non-compatible changes - I totally agree - using the new field with different name AND number.